src/black/__init__.py

```diff
@@ -5174,35 +5174,78 @@ def normalize_numeric_literal(leaf: Leaf) -> None:
     in Python 2 long literals).
     """
     text = leaf.value.lower()
-    if text.startswith(("0o", "0b")):
+    if is_oct_or_binary(text):
         # Leave octal and binary literals alone.
         pass
-    elif text.startswith("0x"):
-        # Change hex literals to upper case.
-        before, after = text[:2], text[2:]
-        text = f"{before}{after.upper()}"
-    elif "e" in text:
-        before, after = text.split("e")
-        sign = ""
-        if after.startswith("-"):
-            after = after[1:]
-            sign = "-"
-        elif after.startswith("+"):
-            after = after[1:]
-        before = format_float_or_int_string(before)
-        text = f"{before}e{sign}{after}"
-    elif text.endswith(("j", "l")):
-        number = text[:-1]
-        suffix = text[-1]
-        # Capitalize in "2L" because "l" looks too similar to "1".
-        if suffix == "l":
-            suffix = "L"
-        text = f"{format_float_or_int_string(number)}{suffix}"
+    elif is_hex(text):
+        text = format_hex(text)
+    elif is_scientific_notation(text):
+        text = format_scientific_notation(text)
+    elif is_long_or_complex_number(text):
+        text = format_long_or_complex_number(text)
     else:
         text = format_float_or_int_string(text)
     leaf.value = text
 
 
+def is_oct_or_binary(text: str) -> bool:
+    """
+    Checks if the supplied string is a number with octal or binary notation
+    """
+    return text.startswith(("0o", "0b"))
+
+
+def is_hex(text: str) -> bool:
+    """Checks if the supplied string is a number with hexadecimal notation"""
+    return text.startswith("0x")
+
+
+def format_hex(text: str) -> str:
+    """
+    Formats a hexidecimal strign like "0x12b3"
+
+    Uses lowercase because of similarity between "B" and "8", which
+    can cause security issues.
+    see: https://github.com/psf/black/issues/1692
+    """
+
+    before, after = text[:2], text[2:]
+    return f"{before}{after.lower()}"
+
+
+def is_scientific_notation(text: str) -> bool:
+    """Checks if the supplied string is a number with scientific notation"""
```

Comment:
This code matches 0xdeadbeef; it's an incorrect abstraction that depends on is_hex and other test having been applied earlier.

Reply:
Valid. Should i move all the checkers into the original function, while still moving the formatting into sub functions? Would that be better.

Reply:
I moved these now :)